File: chap05/fig5-3.R
library(dplyr)
library(ggplot2)

d <- read.csv(file='input/data-shopping-1.csv')
N <- nrow(d)

fit <- readRDS('output/result-model5-3.RDS')
mu_ms <- fit$draws('mu', format='matrix')
N_ms <- nrow(mu_ms)
noise_ms <- t(replicate(N_ms, d$Y)) - mu_ms
# noise_ms <- einsum::einsum('n,m->mn', d$Y, rep(1, N_ms)) - mu_ms

d_est <- data.frame(noise_ms, check.names=FALSE) %>% 
  tidyr::pivot_longer(cols=everything(), names_to='Parameter') %>% 
  mutate(PersonID = readr::parse_number(Parameter))

d_mode <- apply(noise_ms, 2, function(x) {
  dens <- density(x)
  mode_i <- which.max(dens$y)
  mode_x <- dens$x[mode_i]
  mode_y <- dens$y[mode_i]
  c(mode_x, mode_y)
}) %>% 
  t() %>% 
  data.frame() %>% 
  magrittr::set_colnames(c('X', 'Y'))

p <- ggplot() +
  theme_bw(base_size=18) +
  geom_line(data=d_est, aes(x=value, group=PersonID), stat='density', color='black', alpha=0.4) +
  geom_segment(data=d_mode, aes(x=X, xend=X, y=Y, yend=0), color='black', linetype='dashed', alpha=0.4) +
  geom_rug(data=d_mode, aes(x=X), sides='b') +
  labs(x='value', y='density')
ggsave(file='output/fig5-3-left.png', plot=p, dpi=300, w=4, h=3)


s_dens <- density(fit$draws('sigma', format='matrix'))
s_MAP <- s_dens$x[which.max(s_dens$y)]
bw <- 0.01
p <- ggplot(data=d_mode, aes(x=X)) +
  theme_bw(base_size=18) +
  geom_histogram(aes(y=..density..), binwidth=bw, color='black', fill='white') +
  geom_density(color=NA, fill='gray20', alpha=0.5) +
  geom_rug(sides='b') +
  stat_function(fun=function(x) dnorm(x, mean=0, sd=s_MAP), linetype='dashed') +
  labs(x='value', y='density') +
  xlim(range(density(d_mode$X)$x))
ggsave(file='output/fig5-3-right.png', plot=p, dpi=300, w=4, h=3)

File: chap05/fig5-3.py
import pandas
import cmdstanpy
import numpy as np
from scipy import stats
from plotnine import *

d = pandas.read_csv('input/data-shopping-1.csv')
N = len(d)

fit = cmdstanpy.from_csv('output/result-model5-3')
mu_ms = fit.stan_variable(var='mu')
N_ms = len(mu_ms)
noise_ms = np.repeat(d.Y.to_numpy()[None, :], N_ms, axis=0) - mu_ms
# noise_ms = np.einsum('n,m->mn', d.Y, np.repeat(1, N_ms)) - mu_ms

d_est = pandas.DataFrame(noise_ms, columns=['mu[' + str(i) + ']' for i in range(1,N+1)]) 
d_est = pandas.melt(d_est, var_name='Parameter')
d_est['PersonID'] = d_est.Parameter.str.extract('(\d+)')

mode_list = []
for i in range(N):
    density = stats.gaussian_kde(noise_ms[:,i])
    xs = np.linspace(-0.15, 0.15, 201)
    ys = density(xs)
    mode_i = np.argmax(ys)
    mode_x = xs[mode_i]
    mode_y = ys[mode_i]
    mode_list.append([mode_x, mode_y])

d_mode = pandas.DataFrame(mode_list, columns=['X', 'Y'])

p = (ggplot()
    + theme_bw(base_size=18)
    + geom_line(d_est, aes(x='value', group='PersonID'), stat='density', color='black', alpha=0.4)
    + geom_segment(d_mode, aes(x='X', xend='X', y='Y', yend=0), color='black', linetype='dashed', alpha=0.4)
    + geom_rug(d_mode, aes(x='X'), sides='b')
    + labs(x='value', y='density')
) 
p.save(filename='output/fig5-3-left.py.png', dpi=300, width=4, height=3)


s_dens = stats.gaussian_kde(fit.stan_variable(var='sigma'))
xs = np.linspace(-0.15, 0.15, 201)
ys = s_dens(xs)
s_MAP = xs[np.argmax(ys)]
bw = 0.01

p = (ggplot(d_mode, aes(x='X'))
    + theme_bw(base_size=18)
    + geom_histogram(aes(y='..density..'), binwidth=bw, color='black', fill='white')
    + geom_density(fill='gray', alpha=0.5)
    + geom_rug(sides='b')
    + stat_function(fun=stats.norm.pdf, args=dict(loc=0, scale=s_MAP), linetype='dashed')
    + labs(x='value', y='density')
    + xlim(-0.13, 0.13)
) 
p.save(filename='output/fig5-3-right.py.png', dpi=300, width=4, height=3)
